hardinfo/usb_util.c

```diff
@@ -0,0 +1,185 @@
+/*
+ *    HardInfo - Displays System Information
+ *    Copyright (C) 2003-2017 Leandro A. F. Pereira <leandro@hardinfo.org>
+ *    This file
+ *    Copyright (C) 2017 Burt P. <pburt0@gmail.com>
+ *
+ *    This program is free software; you can redistribute it and/or modify
+ *    it under the terms of the GNU General Public License as published by
+ *    the Free Software Foundation, version 2.
+ *
+ *    This program is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU General Public License for more details.
+ *
+ *    You should have received a copy of the GNU General Public License
+ *    along with this program; if not, write to the Free Software
+ *    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#include "hardinfo.h"
+#include "usb_util.h"
+
+usbd *usbd_new() {
+    usbd *s = malloc(sizeof(usbd));
+    if (s) {
+        memset(s, 0, sizeof(usbd));
+    }
+    return s;
+}
+
+void usbd_free(usbd *s) {
+    if (s) {
+        free(s->vendor);
+        free(s->product);
+        free(s->usb_version);
+        free(s->device_version);
+        free(s->dev_class_str);
+        free(s->dev_subclass_str);
+        free(s);
+    }
+}
+
+void usbd_list_free(usbd *s) {
+    usbd *n;
+    while(s != NULL) {
+        n = s->next;
+        usbd_free(s);
+        s = n;
+    }
+}
+
+/* returns number of items after append */
+int usbd_list_append(usbd *l, usbd *n) {
+    int c = 0;
+    while(l != NULL) {
+        c++;
+        if (l->next == NULL) {
+            if (n != NULL) {
+                l->next = n;
+                c++;
+            }
+            break;
+        }
+        l = l->next;
+    }
+    return c;
+}
+
+int usbd_list_count(usbd *s) {
+    return usbd_list_append(s, NULL);
+}
+
+char *_lsusb_lv(char *line, const char *prefix) {
+    if ( g_str_has_prefix(line, prefix) ) {
+        line += strlen(prefix) + 1;
+        return g_strstrip(line);
+    } else
+        return NULL;
+}
+
+int usb_get_device_lsusb(int bus, int dev, usbd *s) {
+    gboolean spawned;
+    gchar *out, *err, *p, *l, *t, *next_nl;
+    gchar *lsusb_cmd = g_strdup_printf("lsusb -s %d:%d -v", bus, dev);
```

Comment:
I wonder if we would benefit from a `h_popen_printf()` that worked pretty much like a `printf()` function. Maybe with `fopencookie()`, this could be used like `popen()`:

```C
    FILE *lsusb = h_popen_printf("lsusb -s %d:%d -v", bus, dev);
    char buffer[1024];

    if (!lsusb)
        return -errno;

    while (fgets(buffer, sizeof(buffer), lsusb)) {
         ...
    }

    h_popen_close(lsusb);
```

This would simplify a lot all these places where `strchr()` and `strend()` are used inside a loop: these could be moved to one of the cookie functions.

(No need to do this for this patch, just thinking out loud.)